clippy_lints/src/multiple_unsafe_ops_per_block.rs

```diff
@@ -0,0 +1,210 @@
+    /// /// # Safety
+    /// ///
+    /// /// - `ptr` must be 4-byte aligned, point to four consecutive
+    /// ///   initialized bytes, and be valid for reads of 4 bytes.
+    /// /// - The bytes pointed to by `ptr` must represent a valid
+    /// ///   `char` when interpreted in the native byte order.
+    /// fn read_char(ptr: *const u8) -> char {
+    ///     // SAFETY: `ptr` is 4-byte aligned, points to four consecutive
+    ///     // initialized bytes, and is valid for reads of 4 bytes.
+    ///     let int_value = unsafe { *ptr.cast::<u32>() };
+    ///
+    ///     // SAFETY: The caller has guaranteed that the four bytes
+    ///     // pointed to by `bytes` represent a valid `char`.
+    ///     unsafe { char::from_u32_unchecked(int_value) }
+    /// }
+    /// ```
+    #[clippy::version = "1.68.0"]
+    pub MULTIPLE_UNSAFE_OPS_PER_BLOCK,
+    restriction,
+    "more than one unsafe operation per `unsafe` block"
+}
+declare_lint_pass!(MultipleUnsafeOpsPerBlock => [MULTIPLE_UNSAFE_OPS_PER_BLOCK]);
+
+impl<'tcx> LateLintPass<'tcx> for MultipleUnsafeOpsPerBlock {
+    fn check_block(&mut self, cx: &LateContext<'tcx>, block: &'tcx hir::Block<'_>) {
+        let mut unsafe_ops = vec![];
+        collect_block(cx, block, &mut unsafe_ops);
+        if unsafe_ops.len() > 1 {
+            span_lint_and_then(
+                cx,
+                MULTIPLE_UNSAFE_OPS_PER_BLOCK,
+                block.span,
+                &format!(
+                    "this `unsafe` block contains {} unsafe operations, expected only one",
+                    unsafe_ops.len()
+                ),
+                |diag| {
+                    for (msg, span) in unsafe_ops {
+                        diag.span_note(span, msg);
+                    }
+                },
+            );
+        }
+    }
+}
+
+fn collect_block<'tcx>(
+    cx: &LateContext<'tcx>,
+    block: &'tcx hir::Block<'_>,
+    unsafe_ops: &mut Vec<(&'static str, Span)>,
+) {
+    if let BlockCheckMode::UnsafeBlock(_) = block.rules {
+        if let Some(expr) = block.expr {
+            collect_unsafe_exprs(cx, expr, unsafe_ops);
+        }
+        for stmt in block.stmts {
+            match stmt.kind {
+                StmtKind::Local(Local { init, els: r#else, .. }) => {
+                    if let Some(init) = init {
+                        collect_unsafe_exprs(cx, init, unsafe_ops);
+                    }
+                    if let Some(r#else) = r#else {
+                        collect_block(cx, r#else, unsafe_ops);
+                    }
+                },
+                StmtKind::Expr(e) | StmtKind::Semi(e) => collect_unsafe_exprs(cx, e, unsafe_ops),
+                StmtKind::Item(_) => {},
+            }
+        }
+    }
+}
+
+fn collect_unsafe_exprs<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx hir::Expr<'_>,
+    unsafe_ops: &mut Vec<(&'static str, Span)>,
+) {
+    let expr = expr.peel_drop_temps();
+
+    for_each_expr_with_closures(cx, expr, |expr| {
```

Comment:
Can't take `Block` directly here, as I still have one recursive call at the bottom, but I figured I can take `impl Visitable`, so that works. Thanks for the comment!